File: apps/backend/src/bot/action-table.ts
/**
 * 18-action discrete action table matching training/action_table.py.
 * Maps action index → InputState for tank controls.
 */
import type { InputState } from '@tankbet/game-engine/physics';

// [up, down, left, right, fire]
const ACTION_TABLE: readonly [boolean, boolean, boolean, boolean, boolean][] = [
  [false, false, false, false, false], // 0:  idle
  [true,  false, false, false, false], // 1:  forward
  [false, true,  false, false, false], // 2:  backward
  [false, false, true,  false, false], // 3:  rotate left
  [false, false, false, true,  false], // 4:  rotate right
  [true,  false, true,  false, false], // 5:  forward + left
  [true,  false, false, true,  false], // 6:  forward + right
  [false, true,  true,  false, false], // 7:  backward + left
  [false, true,  false, true,  false], // 8:  backward + right
  [false, false, false, false, true],  // 9:  fire
  [true,  false, false, false, true],  // 10: forward + fire
  [false, true,  false, false, true],  // 11: backward + fire
  [false, false, true,  false, true],  // 12: left + fire
  [false, false, false, true,  true],  // 13: right + fire
  [true,  false, true,  false, true],  // 14: forward + left + fire
  [true,  false, false, true,  true],  // 15: forward + right + fire
  [false, true,  true,  false, true],  // 16: backward + left + fire
  [false, true,  false, true,  true],  // 17: backward + right + fire
];

export function decodeAction(actionIdx: number): InputState {
  const entry = ACTION_TABLE[actionIdx];
  if (!entry) {
    return { up: false, down: false, left: false, right: false, fire: false };
  }
  return {
    up: entry[0],
    down: entry[1],
    left: entry[2],
    right: entry[3],
    fire: entry[4],
  };
}

File: training/action_table.py
# Action → InputState mapping
ACTION_TABLE = [
    # (up, down, left, right, fire)
    (False, False, False, False, False),  # 0:  idle
    (True,  False, False, False, False),  # 1:  forward
    (False, True,  False, False, False),  # 2:  backward
    (False, False, True,  False, False),  # 3:  rotate left
    (False, False, False, True,  False),  # 4:  rotate right
    (True,  False, True,  False, False),  # 5:  forward + left
    (True,  False, False, True,  False),  # 6:  forward + right
    (False, True,  True,  False, False),  # 7:  backward + left
    (False, True,  False, True,  False),  # 8:  backward + right
    (False, False, False, False, True),   # 9:  fire
    (True,  False, False, False, True),   # 10: forward + fire
    (False, True,  False, False, True),   # 11: backward + fire
    (False, False, True,  False, True),   # 12: left + fire
    (False, False, False, True,  True),   # 13: right + fire
    (True,  False, True,  False, True),   # 14: forward + left + fire
    (True,  False, False, True,  True),   # 15: forward + right + fire
    (False, True,  True,  False, True),   # 16: backward + left + fire
    (False, True,  False, True,  True),   # 17: backward + right + fire
]


def decode_action(action: int) -> dict:
    """Convert an action index (0-17) to a dict of input flags."""
    up, down, left, right, fire = ACTION_TABLE[action]
    return {"up": up, "down": down, "left": left, "right": right, "fire": fire}
